llvm/lib/Transforms/Vectorize/VPlan.cpp

```diff
@@ -129,11 +126,29 @@ void VPDef::dump() const {
 #endif
 
 VPRecipeBase *VPValue::getDefiningRecipe() {
-  return cast_or_null<VPRecipeBase>(Def);
+  auto *Def = dyn_cast<VPDefValue>(this);
+  if (!Def)
+    return nullptr;
+  return cast<VPRecipeBase>(Def->Def);
 }
 
 const VPRecipeBase *VPValue::getDefiningRecipe() const {
-  return cast_or_null<VPRecipeBase>(Def);
+  auto *Def = dyn_cast<VPDefValue>(this);
+  if (!Def)
+    return nullptr;
+  return cast<VPRecipeBase>(Def->Def);
+}
+
+VPDefValue::VPDefValue(VPDef *Def, Value *UV)
+    : VPValue(VPVDefValueSC, UV, nullptr), Def(Def) {
+  assert(Def && "VPDefValue requires a defining recipe");
+  Def->addDefinedValue(this);
+}
+
+VPDefValue::~VPDefValue() {
+  assert(Users.empty() && "trying to delete a VPValue with remaining users");
+  if (Def)
+    Def->removeDefinedValue(this);
```

Comment:
```suggestion
  assert(Def && "VPDefValue requires a defining recipe");
  Def->removeDefinedValue(this);
```
?

Reply:
currenetly this does not work yet, due to how VPDef deletes its values.

Reply:
Worth a comment?